src/NServiceBus.Core/Transports/Msmq/ReceiveWithTransactionScope.cs

```diff
@@ -19,45 +19,118 @@ public ReceiveWithTransactionScope(TransactionOptions transactionOptions)
 
-                    await onMessage(pushContext).ConfigureAwait(false);
-                }
+                        scope.Complete();
+                        return;
+                    }
+
+                    using (var bodyStream = message.BodyStream)
+                    {
+                        var ambientTransaction = new TransportTransaction();
+                        ambientTransaction.Set(Transaction.Current);
+                        var pushContext = new PushContext(message.Id, headers, bodyStream, ambientTransaction, cancellationTokenSource, new ContextBag());
 
-                if (!cancellationTokenSource.Token.IsCancellationRequested)
+                        await onMessage(pushContext).ConfigureAwait(false);
+                    }
+
+                    if (!cancellationTokenSource.Token.IsCancellationRequested)
+                    {
+                        scope.Complete();
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                if (message == null)
                 {
-                    scope.Complete();
+                    throw;
                 }
+                var numberOfRetries = GetNumberOfRetries(message.Id);
+
+                var context = new ErrorContext(ex,
+                    message.Id,
+                    headers ?? new Dictionary<string, string>(),
+                    numberOfRetries
+                    );
+
+                var recoveryAction = await onError(context).ConfigureAwait(false);
+
+                AddToLRU(message.Id, recoveryAction);
             }
         }
 
+        int GetNumberOfRetries(string messageId)
+        {
+            //get this from the LRU/headers etc
+            return 0;
+        }
+
+        void AddToLRU(string messageId, RecoveryAction recoveryAction)
+        {
+
+        }
+
+        bool PerformRecoverabilityAction(Message message)
+        {
+            //check LRU to see if we need to do anything, FLR == do nothing, SLR + Error == move it and return true
+            return false;
+        }
+
+
         TransactionOptions transactionOptions;
 
         static ILog Logger = LogManager.GetLogger<ReceiveWithTransactionScope>();
     }
+
+    class ErrorContext
+    {
+        public Exception Exception { get; }
+        public string MessageId { get; }
+        public Dictionary<string, string> Headers { get; }
+
+        public  int NumberOfRetries{ get;  }
+
+        public ErrorContext(Exception exception, string messageId, Dictionary<string, string> headers, int numberOfRetries)
+        {
+            Exception = exception;
+            MessageId = messageId;
+            Headers = headers;
+            NumberOfRetries = numberOfRetries;
+        }
+    }
+
+    class RecoveryAction
+    {
+        public bool ImmediateRetry { get; set; }
+        public bool DelayedRetry { get; set; }
+
+        public Dictionary<string,string> AdditionalFaultDetailStuff{ get; set; }
```

Comment:
This is the stuff that the core "requests" the transport to attach to the error. For now this can be added to the headers just like now but in the future we want this to be sent out of band to avoid message size issues.

That said we could deal with that in the core if needed and pass a empty dictionary back. Not sure yet what's the best approach